src/zenml/steps/external_artifact.py

```diff
@@ -167,17 +165,46 @@ def upload_if_necessary(self) -> UUID:
         Returns:
             The artifact ID.
         """
-        artifact_store_id = Client().active_stack.artifact_store.id
+        from zenml.client import Client
+        from zenml.utils import artifact_utils
+
+        return self._testable_upload_if_necessary(
+            Client, artifact_utils, model_config
+        )
+
+    def _testable_upload_if_necessary(
+        self,
+        _client: Type["Client"],
+        _artifact_utils: Any,
+        model_config: Optional["ModelConfig"] = None,
+    ) -> UUID:
+        """Uploads the artifact if necessary.
```

Comment:
I carved it out, cause we had unit tests for it, and it all breaks without mocked Client object otherwise. I can sacrifice unit tests in favor of the old method, but is it worth it?

Reply:
I'll make them integrational tests, maybe. Agree that this design is not the best one, it is just a thing proven to be working well for making testable objects with such wild things like inner imports.

Reply:
Reworked